File: JavaScript/76-Minimum Window Substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
    len_t = t.length, count = 0, start = 0;
    let hmap_t = new Map();
    let res = [0, Number.POSITIVE_INFINITY];
    for (let char of t){
        hmap_t.set(char, (hmap_t.get(char)||0) + 1 );
    }
    
    for (let i=0; i<s.length; i++){
        hmap_t.set(s[i], (hmap_t.get(s[i])||0) - 1);
        
        if (hmap_t.get(s[i]) >= 0){
            count += 1;
        }
        
        while (count === len_t){
            hmap_t.set(s[start], hmap_t.get(s[start]) + 1);
            if (hmap_t.get(s[start]) > 0){
                count -= 1;
            }
            if (res[1] - res[0] > i- start ){
                res = [start, i];
            }
            start+=1;
        }
    }
    
    return (res[1] === Number.POSITIVE_INFINITY)? '': s.substring(res[0], res[1]+1);
};

File: C++/76-Minimum Window Substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> tMap;
        int satisfied = 0;
        for (char c:t){
            tMap[c] += 1;
        }
        
        // for (const auto& kv : tMap) {
        //     cout << "Key: " << kv.first << ", Value: " << kv.second << endl;
        // }
        
        unordered_map<char, int> sMap;
        int left = 0;
        int start = 0, end = s.size();
        int tSize = tMap.size(); // we need this because the number of keys in tMap can change
        
        for (int i=0; i<s.size(); i++){
            sMap[s[i]] += 1;
            if (sMap[s[i]] == tMap[s[i]]){
                satisfied += 1;
            }
            while (satisfied == tSize){
                if (end-start+1 > i-left+1){
                    start = left;
                    end = i;
                }
                sMap[s[left]] -= 1;
                if (sMap[s[left]] < tMap[s[left]]){
                    satisfied -= 1;
                }
                left++;
            }
        }
        
        return (end-start == s.size())?"":s.substr(start,end-start+1);
    }
};